scylla/tests/integration/skip_metadata_optimization.rs

```diff
@@ -0,0 +1,87 @@
+use crate::utils::test_with_3_node_cluster;
+use scylla::transport::session::Session;
+use scylla::SessionBuilder;
+use scylla::{prepared_statement::PreparedStatement, test_utils::unique_keyspace_name};
+use scylla_cql::frame::types;
+use scylla_proxy::{
+    Condition, ProxyError, Reaction, ResponseFrame, ResponseReaction, ShardAwareness, TargetShard,
+    WorkerError,
+};
+use std::sync::Arc;
+
+#[tokio::test]
+#[ntest::timeout(20000)]
+#[cfg(not(scylla_cloud_tests))]
+async fn test_skip_result_metadata() {
+    use scylla_proxy::{ResponseOpcode, ResponseRule};
+
+    const NO_METADATA_FLAG: i32 = 0x0004;
+
+    let res = test_with_3_node_cluster(ShardAwareness::QueryNode, |proxy_uris, translation_map, mut running_proxy| async move {
+        // DB preparation phase
+        let session: Session = SessionBuilder::new()
+            .known_node(proxy_uris[0].as_str())
+            .address_translator(Arc::new(translation_map))
+            .build()
+            .await
+            .unwrap();
+
+        let ks = unique_keyspace_name();
+        session.query(format!("CREATE KEYSPACE IF NOT EXISTS {} WITH REPLICATION = {{'class' : 'NetworkTopologyStrategy', 'replication_factor' : 3}}", ks), &[]).await.unwrap();
+        session.use_keyspace(ks, false).await.unwrap();
+        session
+            .query("CREATE TABLE t (a int primary key, b int, c text)", &[])
+            .await
+            .unwrap();
+        session.query("INSERT INTO t (a, b, c) VALUES (1, 2, 'foo_filter_data')", &[]).await.unwrap();
+
+        let mut prepared = session.prepare("SELECT a, b, c FROM t").await.unwrap();
+
+        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+
+        // We inserted this string to filter responses
+        let body_rows = b"foo_filter_data";
+        for node in running_proxy.running_nodes.iter_mut() {
+            let rule = ResponseRule(
+                Condition::ResponseOpcode(ResponseOpcode::Result).and(Condition::BodyContainsCaseSensitive(Box::new(*body_rows))),
+                ResponseReaction::noop().with_feedback_when_performed(tx.clone())
+            );
+            node.change_response_rules(Some(vec![rule]));
+        }
+
+        async fn test_with_flags_predicate(
+            session: &Session,
+            prepared: &PreparedStatement,
+            rx: &mut tokio::sync::mpsc::UnboundedReceiver<(ResponseFrame, Option<TargetShard>)>,
+            predicate: impl FnOnce(i32) -> bool
+        ) {
+            session.execute(prepared, &[]).await.unwrap();
+
+            let (frame, _shard) = rx.recv().await.unwrap();
+            let mut buf = &*frame.body;
+
+            match types::read_int(&mut buf).unwrap() {
+                0x0002 => (),
+                _ => panic!("Invalid result type"),
+            }
+            let result_metadata_flags = types::read_int(&mut buf).unwrap();
+            assert!(predicate(result_metadata_flags));
```

Comment:
Can we deserialize the response using `scylla-cql` instead of doing it manually here?

Reply:
At first, I wanted to use `scylla_cql::frame::response::result::deser_result_metadata`. The thing is that this function is private.

To use this function in the integration tests (placed in `scylla` crate), we would need to:
- make it `pub` in `scylla-cql`
- re-export it as `pub` in `scylla` - unfortunately the function needs to be visible outside of the crate to use it in the integration tests corresponding to the crate.

I've recognized this issue when working on limiting the `pub` re-exports (I'll probably post PR today). There are some utility functions such as: `scylla_cql::frame::types::read/write_int` which unfortunately have to be public, since they are used in the integration tests in `scylla` crate. The only solution to this is to add `#[doc(hidden)]` macro before the re-export, so it's not displayed in the documentation (or possibly add some mention in the docs that these are some utility functions intended for the internal usage).

I could pub re-export the `deser_result_metadata` function and use it in the test case. WDYT?

Reply:
What about `Response::deserialize` (scylla-cql/src/frame/response/mod.rs)? It is already public as far as I can tell, can we use it here?

Reply:
Ugh, ok. Again, as I noted in https://github.com/scylladb/scylla-rust-driver/pull/925#discussion_r1482806217 it will be possible to solve this after #462 because I'm going to change `Response::deserialize` so that it keeps the row data in serialized form.

For now we can keep it as it is, but please add a FIXME comment.